Image2image - swift

littleowl opened the pull request:
**Adds image2image functionality.**

This is only the swift portion. A separate library that generates the model for the VAE Encoder exists here:  https://github.com/apple/ml-stable-diffusion/pull/115

The original PR with both Swift and Python libs can be found here: https://github.com/apple/ml-stable-diffusion/pull/73

Sadly, I could not complete the request for the python simplification. I believe it is better not to do so. Consequently, sorry for the delay splitting up the PR. There are not many changes from the previous PR apart from removing the python and rebasing the project.

In Swift, an Encoder class is created. Various changes to the scheduler, pipeline, and CLI to support input image and strength. CGImage creation from MLShapedArray is moved into its own file along with the new function to create a MLShapedArray from a CGImage. Image loading and preparation is currently handled / optimized with vImage.

This should work with both Schedulers that we have so far. (Thanks @pcuenca) for the fix to work with DPMSolverMultistepScheduler.

Do not erase the below when submitting your pull request:
#########

- [x] I agree to the terms outlined in CONTRIBUTING.md

atiorh commented:
Thank you @littleowl ! I reviewed #115 and leaving the noise situation as is should be fine, please see my comment there :) I believe the only implication of my review on #115 on this PR is that the model I/O names should be changed from camelCase to underscore and that is it. Thank you again for your amazing contributions!

alejandro-isaza commented on swift/StableDiffusion/pipeline/DPMSolverMultistepScheduler.swift:
```diff
@@ -178,5 +178,5 @@ public final class DPMSolverMultistepScheduler: Scheduler {
         
         return prevSample
-    }
+    }    
```
Please revert this.

alejandro-isaza commented on swift/StableDiffusion/pipeline/Scheduler.swift:
```diff
@@ -10,10 +10,13 @@ public protocol Scheduler {
     /// Number of inference steps to be performed
     var inferenceStepCount: Int { get }
-
+    
```
nit. please revert

alejandro-isaza commented on swift/StableDiffusion/pipeline/StableDiffusionPipeline+Resources.swift:
```diff
@@ -74,11 +76,22 @@ public extension StableDiffusionPipeline {
+        // Optional Image Encoder
+        let encoder: Encoder?
+        if
+            let encoderModel = try? MLModel(contentsOf: urls.encoderURL, configuration: config)
```
It would be good to distinguish between no model (missing file) and incorrect model (other error thrown).

littleowl replied:
I did add a new error at first, but now that we implement the `ManagedMLModel` we now just test to see if there is an encoder model file before attempting to load it.

alejandro-isaza commented on swift/StableDiffusion/pipeline/StableDiffusionPipeline.swift:
```diff
@@ -114,12 +123,14 @@ public struct StableDiffusionPipeline: ResourceManaging {
         disableSafety: Bool = false,
-        scheduler: StableDiffusionScheduler = .pndmScheduler,
+        scheduler schedulerType: StableDiffusionScheduler = .pndmScheduler,
         progressHandler: (Progress) -> Bool = { _ in true }
```
This method has too many parameters. We should consider using a `Configuration` struct.

littleowl replied:
Created a struct called `SampleIn`

alejandro-isaza commented on swift/StableDiffusionCLI/main.swift:
```diff
@@ -32,6 +32,12 @@ struct StableDiffusionSample: ParsableCommand {
     var resourcePath: String = "./"
+    
+    @Option(help: "Path to starting image.")
+    var image: String = "none"
```
I think this be optional instead of using "none"

alejandro-isaza reviewed (commented):
Looks good, just a couple of small things.

atiorh commented on swift/StableDiffusion/pipeline/Encoder.swift:
```diff
@@ -0,0 +1,79 @@
+    }
+    
+    /// VAE encoder model + post math and adding noise from schedular
+    var model: MLModel
```
Could you please adopt the [ManagedMLModel](https://github.com/apple/ml-stable-diffusion/blob/main/swift/StableDiffusion/pipeline/ManagedMLModel.swift#L11) class so this model can also be offloaded when required?

littleowl commented:
@atiorh, @alejandro-isaza - Thanks for your reviews! I believe that I resolved all the comments. Please let me know if there is anything more that I can do.

alejandro-isaza commented on swift/StableDiffusion/pipeline/StableDiffusionPipeline+SampleInput.swift:
```diff
@@ -0,0 +1,82 @@
+//  
+//
+//  Created by Timothy Kautz on 1/30/23.
+//
```
Please use the standard header like other files.

alejandro-isaza commented on swift/StableDiffusion/pipeline/StableDiffusionPipeline+SampleInput.swift:
```diff
@@ -0,0 +1,82 @@
+    }
+    
+    /// Image generation configuration
+    public struct SampleInput: Hashable {
```
```suggestion
    public struct Configuration: Hashable {
```

alejandro-isaza commented on swift/StableDiffusion/pipeline/StableDiffusionPipeline+SampleInput.swift:
```diff
@@ -0,0 +1,82 @@
+            seed: UInt32 = 0,
+            guidanceScale: Float = 7.5,
+            disableSafety: Bool = false,
+            schedulerType: StableDiffusionScheduler = .pndmScheduler
```
I don't think we need this initializer, we can just have an initializer that takes `prompt`, everything else can be set after as needed.

alejandro-isaza reviewed (commented):
The header comment needs to be fixed. Other than that just a couple more nits.

littleowl commented:
@alejandro-isaza, thanks again for the review. The additional comments have been resolved.

atiorh commented:
@littleowl  In addition to the file name change that @alejandro-isaza  asked for, if you could just add a quick note in the [Example CLI Usage](https://github.com/apple/ml-stable-diffusion#example-cli-usage) section to indicate that an image argument is now supported, that would be great!

atiorh commented:
I just tested the CLI with different images and it seems that we need to guard against or fix a few things:
1-) When the `--image` file's resolution does not match the VAEEncoder input resolution, we get `zsh: trace trap`
2-) Loading a PNG image with the correct resolution works but trying to load the same image encoded as JPEG fails without a sufficient explanation

atiorh reviewed (commented):
@littleowl I just realized that we left scattered comments after the initial review, apologies! Here are the two remaining items before  we can merge:
- `swift/StableDiffusion/pipeline/StableDiffusionPipeline+SampleInput.swift` renamed to `swift/StableDiffusion/pipeline/StableDiffusionPipeline.Configuration.swift`
- specific error message when input image file format and/or resolution do not match expectations

pj4533 commented:
I know you all have this awesome PR under control, but just wanted to say that I successfully converted the v2 base model VAEEncoder using the main (w/ the python image2image PR), and then converted the rest of v2 base using this PR, and am currently generating image2image using my own swift code. So from an end-user perspective: well done!

saiedg commented:
@pj4533 well done! I'm so jealous! how can I try??

littleowl commented:
Again, sorry for the delay @atiorh.
I added some error detection. It needs to happen right before prediction due to the dynamic model.
I fixed a hard coded instance of 512.
Renamed the file.
Fixed Jpeg Support by utilizing Cocoa.

saiedg commented:
@littleowl congratulations on merging! I'm sooo excited to use it but don't know how. Can you please add instructions in the readme or make a YouTube video?

ynagatomo commented:
here is a sample;
<img width="1679" alt="Screenshot 2023-02-10 at 15 50 52" src="https://user-images.githubusercontent.com/66309582/218036408-ea77b5b1-ca83-4d15-ba65-a053b47959cd.png">

martinlexow commented:
Thanks for your exceptional work @littleowl and thank you for sharing the example @ynagatomo! 

Can you please explain what the `strength` parameter does in this context? It seems to just decrease the steps in my observation (like: steps × strength) and it also doesn’t seem obvious to me why `strength` has to be less than 1.0 to trigger the `imageToImage` mode.

ynagatomo commented:
an experiment on strength;
<img width="640" alt="i2i_1280" src="https://user-images.githubusercontent.com/66309582/218245230-cfd3710a-0aed-445e-8728-8868bc22ac2d.png">
(code: https://github.com/ynagatomo/ImgGenSD2)
